core/src/main/scala/quasar/destination/gbq/GBQDestinationModule.scala

```diff
@@ -0,0 +1,180 @@
+import cats.data.EitherT
+import cats.implicits._
+
+import eu.timepit.refined.auto._
+
+import java.net.{InetSocketAddress, ProxySelector}
+import java.net.Proxy
+import java.net.Proxy.{Type => ProxyType}
+
+import org.asynchttpclient.proxy.{ProxyServer, ProxyServerSelector}
+import org.asynchttpclient.{AsyncHttpClientConfig, DefaultAsyncHttpClientConfig}
+import org.asynchttpclient.uri.{Uri => AUri}
+import org.http4s.{
+  AuthScheme,
+  Credentials,
+  Method,
+  Request,
+  Status,
+  Uri}
+import org.http4s.headers.Authorization
+import org.http4s.client.Client
+import org.http4s.client.asynchttpclient.AsyncHttpClient
+import org.http4s.util.threads.threadFactory
+
+import org.slf4s.Logging
+
+import scala.util.Either
+import scala.collection.JavaConverters._
+
+import scalaz.NonEmptyList
+
+object GBQDestinationModule extends DestinationModule with Logging {
+
+  val Redacted: String = "<REDACTED>"
+
+  def destinationType = DestinationType("gbq", 1L)
+
+  def sanitizeDestinationConfig(config: Json): Json = {
+    config.as[GBQConfig].toOption match {
+      case Some(c) => c.copy(authCfg = Redacted).asJson
+      case _ => jString(Redacted)
+    }
+  }
+
+  def destination[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](
+      config: Json)
+      : Resource[F,Either[InitializationError[Json],Destination[F]]] = {
+
+    val sanitizedConfig: Json = sanitizeDestinationConfig(config)
+    val configOrError = config.as[GBQConfig].toEither.leftMap {
+      case (err, _) =>
+        DestinationError.malformedConfiguration((destinationType, jString(Redacted), err))
+    }
+
+    val init = for {
+      cfg <- EitherT(Resource.pure[F, Either[InitializationError[Json], GBQConfig]](configOrError))
+      client <- EitherT(mkClient.map(_.asRight[InitializationError[Json]]))
+      _ <- EitherT(Resource.liftF(isLive(client, cfg, sanitizedConfig)))
+    } yield new GBQDestination[F](client, cfg, sanitizedConfig): Destination[F]
+    
+    init.value
+  }
+
+  private def mkConfig[F[_]](proxySelector: ProxySelector): AsyncHttpClientConfig =
+    new DefaultAsyncHttpClientConfig.Builder()
+      .setMaxConnectionsPerHost(200)
+      .setMaxConnections(400)
+      .setRequestTimeout(Int.MaxValue)
+      .setReadTimeout(Int.MaxValue)
+      .setConnectTimeout(Int.MaxValue)
+      .setProxyServerSelector(ProxyVoleProxyServerSelector(proxySelector))
+      .setThreadFactory(threadFactory(name = { i =>
+        s"http4s-async-http-client-worker-${i}"
+      })).build
+
+  private def mkClient[F[_]: ConcurrentEffect]: Resource[F, Client[F]] = {
+    Resource.liftF(Search[F]).flatMap(selector =>
+      AsyncHttpClient.resource(mkConfig(selector)))
+  }
+
+  private def isLive[F[_]: Concurrent: ContextShift](
+      client: Client[F],
+      config: GBQConfig,
+      sanitizedConfig: Json)
+      : F[Either[InitializationError[Json], Unit]] = {
+
+    val req = for {
+      accessToken <- GBQAccessToken.token(config.authCfg.getBytes("UTF-8"))
```

Comment:
This code to get the `authCfg` bytes is repeated a few times. I would rather see `GbqConfig` contain `Array[Byte]` for that field and you get the bytes during the decoding, rather than replicating it everywhere.

Reply:
It looks like it would be cleaner to do it as suggested here. @nicflores would that work?

Reply:
I factored out `getBytes` to a single location within `GBQConfig`.